src/BCEngineMCMC.cxx

```diff
@@ -994,24 +976,89 @@ bool BCEngineMCMC::LoadParametersFromTree(TTree* partree, bool loadObservables)
         Par.FillHistograms(p_fill_1d, p_fill_2d);
         Par.SetNbins(p_nbins);
         AddParameter(Par);
-        ++i;
     }
 
     // load user-defined observables
-    if (!loadObservables)
-        return true;
-    fObservables = BCObservableSet();
-    i = 0;
-    while (partree->GetEntryNumberWithIndex(0, i) >= 0) {
-        partree->GetEntryWithIndex(0, i);
-        BCObservable Obs(p_name, p_lowerlimit, p_upperlimit, p_latexname);
-        Obs.SetPrecision(p_precision);
-        Obs.FillHistograms(p_fill_1d, p_fill_2d);
-        Obs.SetNbins(p_nbins);
-        AddObservable(Obs);
-        ++i;
+    if (loadObservables) {
+        fObservables = BCObservableSet();
+        // loop over entries with "parameter" branch = false
+        for (unsigned i = 0; partree->GetEntryNumberWithIndex(0, i) >= 0; ++i) {
+            partree->GetEntryWithIndex(0, i);
+            BCObservable Obs(p_name, p_lowerlimit, p_upperlimit, p_latexname);
+            Obs.SetPrecision(p_precision);
+            Obs.FillHistograms(p_fill_1d, p_fill_2d);
+            Obs.SetNbins(p_nbins);
+            AddObservable(Obs);
+        }
     }
-    return true;
+
+    partree->ResetBranchAddresses();
+}
+
+// --------------------------------------------------------
+void BCEngineMCMC::LoadMCMCParameters(TTree& partree)
+{
+    if (partree.GetEntries() < 1)
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: tree is empty");
+    if (!partree.GetBranch("parameter"))
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: tree missing parameter branch");
+    if (!partree.GetBranch("index"))
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: tree missing index branch");
+    if (!partree.GetBranch("nchains"))
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: tree missing nchains branch");
+    if (!partree.GetBranch("efficiency_0"))
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: tree missing efficiency_0 branch");
+    if (!partree.GetBranch("scale"))
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: tree missing scale branch");
+
+    partree.ResetBranchAddresses();
+
+    // set number of chains
+    unsigned p_nchains;
+    partree.SetBranchAddress("nchains", &p_nchains);
+    partree.GetEntry(0);
+    if (p_nchains == 0)
+        throw std::runtime_error("BCEngineMCMC::LoadMCMCParameters: no chains in branch");
+    SetNChains(p_nchains);
+
+    double p_efficiency[GetNChains()];
```

Comment:
This and the next line use variable-length arrays. Are they necessary? can't you use `std::vector<double> (GetNChains());`? 
I get these warnings in debug mode. Build failure with warnings is enabled on the manual branch

```
../../src/BCEngineMCMC.cxx: In member function ‘void BCEngineMCMC::LoadMCMCParameters(TTree&)’:
../../src/BCEngineMCMC.cxx:1024:37: warning: ISO C++ forbids variable length array ‘p_efficiency’ [-Wvla]
     double p_efficiency[GetNChains()];
                                     ^
../../src/BCEngineMCMC.cxx:1025:32: warning: ISO C++ forbids variable length array ‘p_scale’ [-Wvla]
     double p_scale[GetNChains()];

```

Reply:
I struggled a lot with the ROOT stuff here, but I think I can get that to work.

Reply:
I struggled a lot with the ROOT stuff here, but I think I can get that to work.